### ISSUE TITLE
search result decode other than utf-8

### ISSUE THREAD

**Issue by vaclavrak** [cannatag/ldap3#301]
Hi, recently I found problem in our LDAP implementation search result responses are not in UTF-8, but windows-1250. In current code  is hardcoded utf-8 (operations/search.py:524)   

```Python
entry_dict['dn'] = response[0][3].decode('utf-8')  
```

in my example the response[0][3] value is like 
```Python 
> b'CN=Rak V\xe1clav'.decode('utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe1 in position 8: invalid continuation byte
``` 
but when I changed it to 
```Python 
> b'CN=Rak V\xe1clav'.decode('windows-1250')
'CN=Rak Václav'
``` 

I made a simple patch, what solve my issue. I'll send you a merge request.

**Comment by cannatag** [comment]
Hi Vaclav,
String in Ldap responses should always be in utf-8 as per the current RFCs. Which server are you connecting to?

Giovanni

**Comment by vaclavrak** [comment]
I'm not server admin so I don't know this exactly. 

So far I know is MS Active Directory  LDAP  implementation on Microsoft Server.

**Comment by cannatag** [comment]
Please check code in dev, I've slightly modified your patch. The config parameter is now called ``RESPONSE_DN_ENCODING`` and can be set to one encoding or a list of encodings. If a list is provided ldap3 tries sequentially until an ecoding doesn't return a UnicodeDecodeError or raise an LDAPDNError exception.

Can you try to search the offending entries and to get the cn attribute back? ldap3 still tries to decode the search response attributes with utf-8 (as per the standard RFC). If you can't get the cn back I will apply a similar parameter to attribute values too.

**Comment by vaclavrak** [comment]
I like the way you update the patch. 

the other thing about CN filter. I did several test and I discovered two things
1. when I code string like:
```Python
>>> u'(CN=Rak Václav)'.encode('windows-1250')
b'(CN=Rak V\xe1clav)'
``` 
this was not suprise that result is *byte*, but this is not accepted by search function. Because python3 have natural encoding UTF-8 I could not use any kind of string 

2. if I put into search function CN parameter in UTF-8 string it doesn't find anyting. I also try to search some name without diacritics and it was sucessfull. 


during my test I found interesting thing in conv.py:43 is "DEFAULT_ENCODING" config parameter what solve it. If I set this into "windows-1250" it search correctly names with diacritics. 

I suggect we could use this "DEFAULT_ENCODING" instead of "RESPONSE_DN_ENCODING" for search result, because is universal. Is that correct? If this is not correct, what about put "DEFAULT_ENCODING" in decode chain.

**Comment by cannatag** [comment]
HI, there are 4 different encoding in the ldap3 library:
- encoding from the server: this should always be utf-8 for strings
- encoding to the server: this should always be utf-8 for strings
- encoding from the user: this is what the DEFAULT_ENCODING tries to guess
- encoding to the output logs: this is 'ascii' by default

With ``RESPONSE_DN_ENCODING`` we set a fifth encoding only for the DN of the entries returned by the server during a search operation.

DEFAULT_ENCODING is used by the library to "guess" the conversion of non-unicode strings, It is set to stdin.encoding if not present then tries sys.getdefaultencoding() and finally is set to 'utf-8'. Auto encoding/escaping of filter string is enabled with the ``auto_encode`` and ``auto_escape`` parameters of the Connection object (they both defaults to True).

I don't think we can extend the use of DEFAULT_ENCODING to DN encoding because this should not be needed. The RESPONSE_DN_ENCODING is a "fix" for an erratic server and should not be automaticcaly applied.

Can you tell me what is the DEFAULT_ENCODING value in your case? Also  can you give me the value returnde by ``sys.stdin.encoding`` and ``sys.getdefaultencoding()``? I'd like to underrstand why you have to set it manually.

Thanks,
Giovanni

**Comment by cannatag** [comment]
The RESPONSE_DN_ENCODING fix will be released in ldap3 v 2.2.2

**Comment by cannatag** [comment]
I've also added a 'raw_dn' key in the search response dictionary. It contains the DN byte value returned by the search.

**Comment by cannatag** [comment]
released in 2.2.2
